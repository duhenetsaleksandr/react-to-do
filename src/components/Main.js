import React, { useState, useContext } from 'react';
import '../styles/main.scss';
import CreateToDo from './CreateToDo';
import ToDoList from './ToDoList';
import Context from '../context';

export default function Main(props) {
    const inputInitialState = '';
    const editModeInitialState = { status: false, todo: {} };
    const { createTodo, editTodo } = useContext(Context);

    const [ inputValue, setInputValue ] = useState(inputInitialState);
    const [ editMode, setEditMode ] = useState(editModeInitialState);

    const activateEditState = (todo) => {
        setInputValue(todo.title);
        setEditMode({ status: true, todo });
    }

    const clearAllStates = () => {
        setInputValue(inputInitialState);
        setEditMode(editModeInitialState);
    };

    const handlerSubmitForm = (event) => {
        event.preventDefault();
        const value = inputValue.trim();
        if (!editMode.status) {
            value && createTodo(value);
        } else {
            value && editTodo(editMode.todo.id, value);
        }
        clearAllStates();
    }

    const handlerChangeInput = (event) => {
        setInputValue(event.target.value);
    }

    return (
        <main className="main">
            <div className="container main__container">
                <div className="todos-wrapper">
                    <CreateToDo
                        editStateInfo={editMode}
                        inputValue={inputValue}
                        handlerChangeInput={handlerChangeInput}
                        handlerSubmitForm={handlerSubmitForm}
                    />
                    <ToDoList
                        todos={props.todos}
                        onEdit={activateEditState}
                    />
                </div>
            </div>
        </main>
    );
}
