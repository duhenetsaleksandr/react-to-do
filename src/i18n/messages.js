import { LOCALES } from './locales'

export const messages = {
    [LOCALES.ENGLISH]: {
        copyright: '©Copyright. All rights deserved.',
        button_add: 'Add task',
        button_edit: 'Edit task',
        placeholder: 'Enter new todo',
        not_todos: 'Not found todos!',
    },
    [LOCALES.RUSSIAN]: {
        copyright: '©Копирайт. Все права защищены.',
        button_add: 'Добавить',
        button_edit: 'Редактировать',
        placeholder: 'Введите новую задачу',
        not_todos: 'Задачи не найдены!',
    },
};
