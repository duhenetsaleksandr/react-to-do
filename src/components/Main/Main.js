import React, { useState } from 'react';
import './Main.scss';
import CreateToDo from 'components/CreateToDo/CreateToDo';
import ToDoList from 'components/ToDoList/ToDoList';

const initialState = {
    editMode: {
        status: false,
        todo: {},
    },
};

const Main = ({ todos, createTodo, deleteTodo, editTodo, toggleTodo }) => {
    const [ state, setState ] = useState(initialState);

    const activateEditState = (todo) => setState({ editMode: { status: true, todo }});

    const clearEditMode = () => {
        setState(initialState);
    };

    return (
        <main className="main">
            <div className="container main__container">
                <div className="todos-wrapper">
                    <CreateToDo
                        editStateInfo={state.editMode}
                        createTodo={createTodo}
                        editTodo={editTodo}
                        clearEditMode={clearEditMode}
                    />
                    <ToDoList
                        todos={todos}
                        toggleTodo={toggleTodo}
                        deleteTodo={deleteTodo}
                        onEdit={activateEditState}
                    />
                </div>
            </div>
        </main>
    );
}

export default Main;
