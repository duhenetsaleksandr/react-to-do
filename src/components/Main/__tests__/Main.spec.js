import React from 'react';
import { act } from '@testing-library/react';
import Main from '../Main';
import { IntlProvider } from 'react-intl';
import { messages } from 'i18n/messages';
import { LOCALES } from 'i18n/locales';

const todos = [
    {
        id: 1,
        title: 'create app',
        completed: false
    },
    {
        id: 2,
        title: 'create app one more time',
        completed: true
    }
];

const props = {
    todos,
    createTodo: jest.fn(),
    deleteTodo: jest.fn(),
    editTodo: jest.fn(),
    toggleTodo: jest.fn(),
};

const initialState = {
    editMode: {
        status: false,
        todo: {},
    },
};

describe('should render Main component', () => {
    let component;
    beforeEach(() => {
        component = mount(<Main {...props} />, {
            wrappingComponent: IntlProvider,
            wrappingComponentProps: {
                messages: messages[LOCALES.ENGLISH],
                locale: LOCALES.ENGLISH,
                defaultLocale: LOCALES.ENGLISH,
            },
        });
    });

    it('should contain tag main with class .main', () => {
        const main = component.find('main.main');
        expect(main).toHaveLength(1);
    });

    it('should contain div with class .container and .main__container', () => {
        const mainContainer = component.find('div.container.main__container');
        expect(mainContainer).toHaveLength(1);
    });

    it('should contain div with class .todos-wrapper', () => {
        const todosWrapper = component.find('div.todos-wrapper');
        expect(todosWrapper).toHaveLength(1);
    });

    it('should contain CreateToDo component', () => {
        const createTodo = component.find('CreateToDo');
        expect(createTodo).toHaveLength(1);
    });

    it('should contain ToDoList component', () => {
        const toDoList = component.find('ToDoList');
        expect(toDoList).toHaveLength(1);
    });

    it('should props editStateInfo in CreateToDo component equal initial state editMode', () => {
        const createTodo = component.find('CreateToDo');
        expect(createTodo.getElement().props.editStateInfo).toEqual(initialState.editMode);
    });

    it('should props editStateInfo in CreateToDo component equal state editMode true', () => {
        act(() => {
            component.find('ToDoList').getElement().props.onEdit(todos[0]);
        });
        component.update();
        expect(component.find('CreateToDo').getElement().props.editStateInfo).toEqual({ status: true, todo: todos[0] });
    });

    it('should reset state to initialState', () => {
        act(() => {
            component.find('ToDoList').getElement().props.onEdit(todos[0]);
        });
        component.update();
        expect(component.find('CreateToDo').getElement().props.editStateInfo).toEqual({ status: true, todo: todos[0] });
        act(() => {
            component.find('CreateToDo').getElement().props.clearEditMode();
        });
        component.update();
        expect(component.find('CreateToDo').getElement().props.editStateInfo).toEqual(initialState.editMode);
    });

    it('should render Main component snapshot', () => {
        const component = shallow(<Main {...props} />);
        expect(component).toMatchSnapshot();
    });
});
