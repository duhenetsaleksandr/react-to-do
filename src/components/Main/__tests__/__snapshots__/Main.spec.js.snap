// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Main component should render Main component snapshot 1`] = `
<main
  className="main"
>
  <div
    className="container main__container"
  >
    <div
      className="todos-wrapper"
    >
      <CreateToDo
        clearEditMode={[Function]}
        createTodo={[MockFunction]}
        editStateInfo={
          Object {
            "status": false,
            "todo": Object {},
          }
        }
        editTodo={[MockFunction]}
      />
      <ToDoList
        deleteTodo={[MockFunction]}
        onEdit={[Function]}
        todos={
          Array [
            Object {
              "completed": false,
              "id": 1,
              "title": "create app",
            },
            Object {
              "completed": true,
              "id": 2,
              "title": "create app one more time",
            },
          ]
        }
        toggleTodo={[MockFunction]}
      />
    </div>
  </div>
</main>
`;
